Core: Typescript types when accessing nested properties don't consider the undefined case

```
function getUserIdsAsEids(): ORTBRequest['user']['eids'] {
  return getEids(initializedSubmodules.combined)
}
```
Since `ORTBRequest['user']` is either a `User` or `undefined`, there is a problem accessing `eids` because that property doesn't exist on `undefined`.

The correct way would be in this case to use

```
NonNullable<ORTBRequest['user']>['eids']
```
as return type. And since that results in `EID[] | undefined`, I think a closer match with the code (which will always return an array and never `undefined`) is

```
NonNullable<NonNullable<ORTBRequest['user']>['eids']>
```

I didn't yet create a PR for this, as I see this pattern of omitting the `NonNullable` part in quite a few places.
I haven't figured out yet why it isn't a problem in the Prebid codebase. I assume something in either the linter or the `tsconfig.json` allows this, and I don't really know what or why.

But at the moment, in my own project, I have to use:
```
const userIdsAsEids: NonNullable<NonNullable<prebid.ORTBRequest['user']>['eids']> = pbjs.getUserIdsAsEids();
```
If I don't specify a type, the TS compiler uses `any` as type for `userIdsAsEids`

Could you share more about your setup - tsconfig.json, maybe typescript version? 

When I try this minimal example:

```typescript
import pbjs from 'prebid.js';
import 'prebid.js/modules/userId';

pbjs.getUserIdsAsEids()[0].test;
```

with tsconfig.json

```
{
  "compilerOptions": {
    "module": "NodeNext",   
    "skipLibCheck": true,
  }
}

```

`tsc` does what I'd expect:

```
src/index.ts:4:28 - error TS2339: Property 'test' does not exist on type 'EID'.

4 pbjs.getUserIdsAsEids()[0].test;
```

--- a/gulp.precompilation.js
+++ b/gulp.precompilation.js
@@ -227,13 +227,15 @@ function precompile(options = {}) {
       generateModuleSummary,
       generateGlobalDef(options),
     ]),
-    helpers.execaTask('tsc -p tsconfig-strict.json')
-  ]);
+  ].concat(options.dev ? [] : [
+    'ts-strict'
+  ]));
 }
 
 
 gulp.task('ts', helpers.execaTask('tsc'));
-gulp.task('ts-dev', helpers.execaTask('tsc --incremental'))
+gulp.task('ts-dev', helpers.execaTask('tsc --incremental'));
+gulp.task('ts-strict', helpers.execaTask('tsc -p tsconfig-strict.json'));
 gulp.task('transpile', babelPrecomp());
 gulp.task('precompile-dev', precompile({dev: true}));
 gulp.task('precompile', precompile());

--- a/gulp.precompilation.js
+++ b/gulp.precompilation.js
@@ -226,7 +226,8 @@ function precompile(options = {}) {
       generateCoreSummary,
       generateModuleSummary,
       generateGlobalDef(options),
-    ])
+    ]),
+    helpers.execaTask('tsc -p tsconfig-strict.json')
   ]);
 }
 

--- a/libraries/ortbConverter/converter.ts
+++ b/libraries/ortbConverter/converter.ts
@@ -65,9 +65,9 @@ type Params<B extends BidderCode> = {
     }
   ) => ORTBRequest;
   [BID_RESPONSE]: (
-    bid: ORTBResponse['seatbid'][number]['bid'][number],
+    bid: NonNullable<ORTBResponse['seatbid']>[number]['bid'][number],
     context: Context & {
-      seatbid: ORTBResponse['seatbid'][number];
+      seatbid: NonNullable<ORTBResponse['seatbid']>[number];
       imp: ORTBImp;
       bidRequest: BidRequest<B>;
       ortbRequest: ORTBRequest;
@@ -97,7 +97,7 @@ type Customizers<B extends BidderCode> = {
 
 type Overrides<B extends BidderCode> = {
   [M in keyof Params<B>]?: {
-    [name: string]: (orig: Processors<B>[M][string], ...args: Parameters<Processors<B>[M][string]>) => void;
+    [name: string]: (orig: NonNullable<Processors<B>[M]>[string], ...args: Parameters<NonNullable<Processors<B>[M]>[string]>) => void;
   }
 }
 

--- a/libraries/vidazooUtils/vidazooTypes.ts
+++ b/libraries/vidazooUtils/vidazooTypes.ts
@@ -29,7 +29,8 @@ export interface VidazooBaseBidderParams {
  * Bid floor value.
  */
 export type Ext = {
-  [key: string]: Record<string, any>;
+  [key: string]: Record<string, any>
+} & {
   customParameters?: CustomParameters;
 }
 

--- a/modules/userId/index.ts
+++ b/modules/userId/index.ts
@@ -45,7 +45,7 @@ import { activityParams } from '../../src/activities/activityParams.js';
 import { USERSYNC_DEFAULT_CONFIG, type UserSyncConfig } from '../../src/userSync.js';
 import type { ORTBRequest } from "../../src/types/ortb/request.d.ts";
 import type { AnyFunction, Wraps } from "../../src/types/functions.d.ts";
-import type { ProviderParams, UserId, UserIdProvider, UserIdConfig, IdProviderSpec, ProviderResponse } from "./spec.ts";
+import type { EID, ProviderParams, UserId, UserIdProvider, UserIdConfig, IdProviderSpec, ProviderResponse } from "./spec.ts";
 import {
   ACTIVITY_PARAM_COMPONENT_NAME,
   ACTIVITY_PARAM_COMPONENT_TYPE,
@@ -469,7 +469,7 @@ export function enrichEids(ortb2Fragments) {
 
 declare module '../../src/adapterManager' {
   interface BaseBidRequest {
-    userIdAsEids: ORTBRequest['user']['eids'];
+    userIdAsEids?: EID[];
   }
 }
 
@@ -676,7 +676,7 @@ function getUserIds() {
  * This function will be exposed in global-name-space so that userIds stored by Prebid UserId module can be used by external codes as well.
  * Simple use case will be passing these UserIds to A9 wrapper solution
  */
-function getUserIdsAsEids(): ORTBRequest['user']['eids'] {
+function getUserIdsAsEids(): EID[] {
   return getEids(initializedSubmodules.combined)
 }
 
@@ -685,7 +685,7 @@ function getUserIdsAsEids(): ORTBRequest['user']['eids'] {
  * Simple use case will be passing these UserIds to A9 wrapper solution
  */
 
-function getUserIdsAsEidBySource(sourceName: string): ORTBRequest['user']['eids'][0] | undefined {
+function getUserIdsAsEidBySource(sourceName: string): EID | undefined {
   return getUserIdsAsEids().filter(eid => eid.source === sourceName)[0];
 }
 

--- a/modules/userId/index.ts
+++ b/modules/userId/index.ts
@@ -31,7 +31,8 @@ import {
   isPlainObject,
   logError,
   logInfo,
-  logWarn, mergeDeep
+  logWarn,
+  mergeDeep
 } from '../../src/utils.js';
 import { getPPID as coreGetPPID } from '../../src/adserver.js';
 import { defer, delay, PbPromise } from '../../src/utils/promise.js';
@@ -43,9 +44,16 @@ import { isActivityAllowed, registerActivityControl } from '../../src/activities
 import { ACTIVITY_ACCESS_DEVICE, ACTIVITY_ENRICH_EIDS } from '../../src/activities/activities.js';
 import { activityParams } from '../../src/activities/activityParams.js';
 import { USERSYNC_DEFAULT_CONFIG, type UserSyncConfig } from '../../src/userSync.js';
-import type { ORTBRequest } from "../../src/types/ortb/request.d.ts";
 import type { AnyFunction, Wraps } from "../../src/types/functions.d.ts";
-import type { EID, ProviderParams, UserId, UserIdProvider, UserIdConfig, IdProviderSpec, ProviderResponse } from "./spec.ts";
+import type {
+  EID,
+  IdProviderSpec,
+  ProviderParams,
+  ProviderResponse,
+  UserId,
+  UserIdConfig,
+  UserIdProvider
+} from "./spec.ts";
 import {
   ACTIVITY_PARAM_COMPONENT_NAME,
   ACTIVITY_PARAM_COMPONENT_TYPE,

--- a/modules/userId/spec.ts
+++ b/modules/userId/spec.ts
@@ -6,6 +6,8 @@ import type { ORTBRequest } from "../../src/types/ortb/request";
 
 export type UserIdProvider = string;
 
+export type EID = NonNullable<NonNullable<ORTBRequest['user']>['eids']>[number];
+
 export interface UserId {
   [idName: string]: unknown;
 }
@@ -125,7 +127,7 @@ type EIDConfig<K extends keyof UserId> = {
   getUidExt?: (id: IdValue<K>) => Ext;
 }
 
-type EIDFn<K extends keyof UserId, P extends UserIdProvider> = (ids: IdValue<K>[], config: UserIdConfig<P>) => ORTBRequest['user']['eids'] | ORTBRequest['user']['eids'][number];
+type EIDFn<K extends keyof UserId, P extends UserIdProvider> = (ids: IdValue<K>[], config: UserIdConfig<P>) => EID | EID[];
 
 export type IdProviderSpec<P extends UserIdProvider> = StorageDisclosure & {
   /**

--- a/modules/videoModule/index.ts
+++ b/modules/videoModule/index.ts
@@ -147,7 +147,7 @@ export function PbVideo(videoCore_, getConfig_, pbGlobal_, requestBids_, pbEvent
     return videoCore.getOrtbVideo(divId);
   }
 
-  function getOrtbContent(divId: string): DeepPartial<ORTBRequest['site']['content']> {
+  function getOrtbContent(divId: string): DeepPartial<NonNullable<ORTBRequest['site']>['content']> {
     return videoCore.getOrtbContent(divId);
   }
 

--- a/src/adapterManager.ts
+++ b/src/adapterManager.ts
@@ -155,13 +155,13 @@ export interface BaseBidRequest extends ContextIdentifiers, Pick<AdUnit, typeof 
   ortb2: DeepPartial<ORTBRequest>;
 }
 
-export interface StoredBidRequest extends BaseBidRequest, Omit<{ [K in keyof AdUnitBidderBid<BidderCode>]?: undefined }, keyof BaseBidRequest> {
+export interface StoredBidRequest extends BaseBidRequest, Omit<{ [K in keyof AdUnitBidderBid<BidderCode>]?: undefined | null }, keyof BaseBidRequest> {
   bidder: null;
   src: typeof S2S.SRC;
 }
 type BidderBidRequest<BIDDER extends BidderCode> = BaseBidRequest & AdUnitBidderBid<BIDDER>;
 
-export type BidRequest<BIDDER extends (BidderCode | null)> = BIDDER extends null ? StoredBidRequest : BidderBidRequest<BIDDER>;
+export type BidRequest<BIDDER extends (BidderCode | null)> = BIDDER extends null ? StoredBidRequest : BidderBidRequest<NonNullable<BIDDER>>;
 
 export interface BaseBidderRequest<BIDDER extends BidderCode | null> {
   /**
@@ -208,7 +208,7 @@ export interface ClientBidderRequest<BIDDER extends BidderCode> extends BaseBidd
   src: 'client';
 }
 
-export type BidderRequest<BIDDER extends BidderCode | null> = ClientBidderRequest<BIDDER> | S2SBidderRequest<BIDDER>;
+export type BidderRequest<BIDDER extends BidderCode | null> = ClientBidderRequest<NonNullable<BIDDER>> | S2SBidderRequest<BIDDER>;
 
 const ADUNIT_BID_PROPERTIES = [
   'nativeParams',

--- a/src/auction.ts
+++ b/src/auction.ts
@@ -947,7 +947,7 @@ export interface DefaultTargeting {
   /**
    * Advertiser domain.
    */
-  [TARGETING_KEYS.ADOMAIN]: Bid['meta']['advertiserDomains'][0];
+  [TARGETING_KEYS.ADOMAIN]: NonNullable<NonNullable<Bid['meta']>['advertiserDomains']>[0];
   /**
    * Primary category ID.
    */

--- a/src/audio.ts
+++ b/src/audio.ts
@@ -40,7 +40,7 @@ export const ORTB_AUDIO_PARAMS = new Map(ORTB_PARAMS);
 
 export type AudioContext = typeof INSTREAM;
 
-export interface AudioMediaType extends BaseMediaType, Pick<ORTBImp['audio'], (typeof ORTB_PARAMS)[number][0]> {
+export interface AudioMediaType extends BaseMediaType, Pick<NonNullable<ORTBImp['audio']>, (typeof ORTB_PARAMS)[number][0]> {
   context?: AudioContext;
 }
 

--- a/src/banner.ts
+++ b/src/banner.ts
@@ -24,7 +24,7 @@ const ORTB_PARAMS = [
  * reference: https://github.com/InteractiveAdvertisingBureau/openrtb2.x/blob/main/2.6.md
  */
 export const ORTB_BANNER_PARAMS = new Map(ORTB_PARAMS);
-export interface BannerMediaType extends BaseMediaType, Partial<Pick<ORTBImp['banner'], (typeof ORTB_PARAMS)[number][0]>> {
+export interface BannerMediaType extends BaseMediaType, Partial<Pick<NonNullable<ORTBImp['banner']>, (typeof ORTB_PARAMS)[number][0]>> {
   /**
    * All sizes this ad unit can accept.
    * Examples: [400, 600], [[300, 250], [300, 600]].

--- a/src/config.ts
+++ b/src/config.ts
@@ -307,7 +307,7 @@ interface GetConfig {
   (): Config;
     <S extends DeepProperty<Config> | string>(setting: S): S extends DeepProperty<Config> ? TypeOfDeepProperty<Config, S> : unknown;
     (topic: typeof ALL_TOPICS, listener: (config: Config) => void, options?: GetConfigOptions): UnregistrationFn;
-    <S extends DeepProperty<Config> | string>(topic: S, listener: (config: TopicalConfig<S>) => void, options?: GetConfigOptions): UnregistrationFn;
+    <S extends NonNullable<DeepProperty<Config>> | string>(topic: S, listener: (config: TopicalConfig<S>) => void, options?: GetConfigOptions): UnregistrationFn;
     (listener: (config: Config) => void, options?: GetConfigOptions): UnregistrationFn;
 }
 

--- a/src/video.ts
+++ b/src/video.ts
@@ -58,7 +58,7 @@ export const ORTB_VIDEO_PARAMS = new Map(ORTB_PARAMS);
 
 export type VideoContext = typeof INSTREAM | typeof OUTSTREAM | typeof ADPOD;
 
-export interface VideoMediaType extends BaseMediaType, Pick<ORTBImp['video'], (typeof ORTB_PARAMS)[number][0]> {
+export interface VideoMediaType extends BaseMediaType, Pick<NonNullable<ORTBImp['video']>, (typeof ORTB_PARAMS)[number][0]> {
   context: VideoContext;
   playerSize?: Size | Size[];
 }

--- a/tsconfig-strict.json
+++ b/tsconfig-strict.json
@@ -1,0 +1,10 @@
+{
+  "README": "This is used to run the type checker on strict mode on the compiled output, to make sure that it can be consumed in strict mode",
+  "extends": "./tsconfig.json",
+  "compilerOptions": {
+    "strict": true,
+    "noEmit": true,
+  },
+  "include": ["src/types/local/**/*.ts", "./dist/src/**/*.ts"],
+  "exclude": []
+}

